plugin/trino-hive/src/test/java/io/trino/plugin/hive/crypto/TestHiveParquetEncryption.java

```diff
@@ -0,0 +1,189 @@
+     */
+    private java.nio.file.Path warehouseDir;
+
+    @Override
+    protected DistributedQueryRunner createQueryRunner()
+            throws Exception
+    {
+        warehouseDir = Files.createTempDirectory("pme_hive");
+
+        Map<String, String> properties = ImmutableMap.<String, String>builder()
+                .put("hive.metastore", "file")
+                .put("hive.metastore.catalog.dir", warehouseDir.toUri().toString())
+                // disable env retriever; we inject keys via custom module
+                .put("pme.environment-key-retriever.enabled", "false")
+                .buildOrThrow();
+
+        return HiveQueryRunner.builder()
+                .setHiveProperties(properties)
+                .setHiveModule(new TestingParquetEncryptionModule(FOOTER_KEY, COLUMN_KEY))
+                .build();
+    }
+
+    @Test
+    public void testEncryptedParquetRead()
+            throws Exception
+    {
+        // 1) write encrypted file inside warehouse
+        java.nio.file.Path dataDir = Files.createDirectories(warehouseDir.resolve("pme_data"));
+        java.nio.file.Path parquetFile = dataDir.resolve("data.parquet");
+        writeEncryptedFile(parquetFile);
+
+        // 2) create external table
+        String location = Location.of(String.valueOf(dataDir.toUri())).toString();
+        assertUpdate("""
+                CREATE TABLE enc_age(age INT)
+                WITH (external_location = '%s', format = 'PARQUET')
+                """.formatted(location));
+
+        // 3) verify results
+        MaterializedResult result = computeActual("SELECT COUNT(*), MIN(age), MAX(age) FROM enc_age");
+        assertThat(result.getMaterializedRows().get(0).getField(0)).isEqualTo(100L);                    // count
+        assertThat(result.getMaterializedRows().get(0).getField(1)).isEqualTo(0);   // min
+        assertThat(result.getMaterializedRows().get(0).getField(2)).isEqualTo(99);  // max
+    }
+
+    // ───────────────────────── writer ─────────────────────────
+    private static void writeEncryptedFile(java.nio.file.Path path)
+            throws Exception
+    {
+        var schema = MessageTypeParser.parseMessageType(
+                "message doc { required int32 age; }");
+
+        /*
+         *  ─────────────────────────  IMPORTANT  ─────────────────────────
+         *
+         *  **This unit-test reuses one synthetic 128-bit key for every file.**
+         *
+         *  That is fine for *tests*, but **DO NOT** follow this pattern in
+         *  production: AES-GCM/CTR has a hard upper-bound (~2³² blocks) on the
+         *  number of encryption operations that may be performed with the same
+         *  key/non-repeating-nonce combination. Exceeding that limit
+         *  catastrophically weakens the cipher.
+         *
+         *  In real deployments you should:
+         *    • Generate a fresh data-encryption-key (DEK) per file *or* per
+         *      small batch of files, **or**
```

Comment:
I'm not sure about the second **or**. If a fresh DEK is generated per file or batch of files, the key-id → DEK-bytes map will let readers understand what DEK was used in a file.

Reply:
Also, this is one example of DEK management. Another popular approach is envelope encryption with KMS.